tensorflow_model_optimization/g3doc/guide/clustering/clustering_comprehensive_guide.ipynb

```diff
@@ -0,0 +1,536 @@
+{
```

Comment:
<p>1. "Generally better to start from pre-trained weights of a model which has acceptable accuracy"</p><p> </p><p>Generally better in contrast to what? (having no pre-trained weights or pre-trained weights with poor accuracy). It's not clear to a reader. Having pre-trained weights with acceptable accuracy as opposed to poor accuracy seems somewhat obvious. </p><p> </p><p>i'd keep the phrasing consistent with what's done for the "Cluster a whole model".</p><p> </p><p>2. Freeze most of the earlier layers before the layers to be clustered during fine-tuning </p><p> </p><p>Why not freeze all layers, if it's important? I'd also be more specific if applicable (e.g. freeze the weights of most of the earlier layers). </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/model-optimization/pull/330/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Reply:
<ol><li>yes. I agree.</li><li>number of frozen layers is a hyperparameter to tune and it varies for different models. Found it through the experiments: depends on the layer we cluster, sometime it has lots of redundancy so that clustering will not significantly affect layer representations of itself or the layers before it. However, for other cases, it might not apply.</li></ol>

Reply:
2/3. Ok thanks. Followup questions if known:
a) How much did freezing benefit accuracy in the experiments? I didn't see you need to use this for the Official Models mobilenet v1 example.
b) Is there a known pattern to these freezing? Something like: freeze the first N layers before the layers to be clustered where N includes most of the layers (e.g. the analogy to transfer learning). 

I'm assuming there isn't a known pattern to the freezing. I'd clarify with the details you said. 

Example: Treat the freezing of layers prior to the layers to be clustered as a hyperparameter. Empirically, freezing most layers is ideal.

Reply:
<blockquote>3. There might be a way of evaluate a layer's redundancy but I have not got time to look into it yet.</blockquote><p>No idea whether people have already seen this, but I did a quick search and found this interesting <a href="https://arxiv.org/pdf/1912.08881.pdf" target="_blank">paper</a>, especially the line 'Since relevance scores are computed for every layer and neuron from the model output to the input, these relevance scores essentially reflect the importance of every single element of a model and its contribution to the information flow through the network – a natural candidate to be used as pruning criterion'. The paper is for pruning but if the metric has linear computation cost it might be useful for model optimization in general. I will think about it in the background of other work.</p>

Reply:
<p>a) and b) are addressed in an email 'on empirical results for Jupyter notebooks' I sent last Friday.</p>